### ISSUE TITLE
ExpandableStringExpressionAst doesn't expose sufficient data to round-trip

### ISSUE THREAD

**Issue by redoz** [PowerShell/PowerShell#5365]
Given an ```ExpandableStringExpressionAst``` node it is not possible to reconstruct the original code. The required data is only hidden away in the internal ```FormatExpression``` property.

Steps to reproduce
------------------

```powershell
{ "`$foo = $foo" }.Ast.EndBlock.Statements[0].PipelineElements[0].Expression
```

Actual behavior
---------------

```none
Value              : $foo = $foo
StringConstantType : DoubleQuoted
NestedExpressions  : {$foo}
StaticType         : System.String
Extent             : "`$foo = $foo"
Parent             : "`$foo = $foo"
```

```NestedExpressions[0]``` is of type ```VariableExpressionAst``` which also has no information (beyond ```Extent``` to indicate where it came from.

With the properties exposed in the ast node there is no way of knowing which instance of ```$foo``` should be expanded (```Extent``` doesn't count)

Environment data
----------------

```powershell
Name                           Value
----                           -----
PSVersion                      6.0.0-beta.9
PSEdition                      Core
GitCommitId                    v6.0.0-beta.9
OS                             Microsoft Windows 10.0.16299
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

For what it's worth this also applies to PowerShell 5.

**Comment by lzybkr** [comment]
Are you saying the extent of NestedExpressions doesn't count?

Not exposing `FormatExpression` is intentional - that should be considered an implementation detail, e.g. if some other formatting engine was used.

**Comment by redoz** [comment]
@lzybkr The Extent (as far as I could tell) is a "this is where this came from in the source code" which doesn't seem like something you'd want to rely on when trying to interpret the AST.

**Comment by lzybkr** [comment]
You're right, and many uses of Extent are wrong - I see this all the time in so many different tools, but in this case, I think it's reasonable.

It might help to understand what you're trying to do so I can offer the best guidance or if the api needs changes to enable your scenario.

**Comment by redoz** [comment]
Well it started out as an experiment of rewriting powershell through AST mutations, but then I realized the Extent property is static and doesn't reflect the actual tree. So I created a code generator/serializer: https://gist.github.com/redoz/659cbbcf2db1ce11cb6032f261d76d05

I test it by parsing powershell files, serializing the AST back to code, parsing that code, then comparing the ASTs recursively. But I got stuck on the ```ExpandableStringExpressionAst``` node, I looked at the source and couldn't figure out how it even worked until I noticed the internal ```FormatExpression``` with accompanying constructors.

**Comment by lzybkr** [comment]
It is a bit of a weak spot. I wrote an [instrumenting profiler](https://code.msdn.microsoft.com/Script-Line-Profiler-Sample-80380291) when first designing the Ast to make sure it was possible to generate a new Ast using only public apis and that exercise pointed out that expandable strings weren't possible at the time. I added the public constructor which was necessary, but maybe not sufficient.

I still think use of the `Extent` of each `NestedExpression` is appropriate - as in, if we were to provide a more friendly api, I that's probably how it would work.

The bottom line - I'm open to suggestions for new apis, but need a lot more convincing if the api exposes an implementation detail like `FormatExpression`.

**Comment by redoz** [comment]
It might be an idea to simply split it into segments so "`$foo = $foo" contains two expressions, one constant string for "$foo = " and one variable reference for $foo, this would change the semantics but not the structure of the API. You would need a new constructor to clone it though, or make sure the backticks would be preserved in the ```Value``` property (as it is when using the current public constructor).

Or the ```NestedExpressions``` property could be a list of wrappers, something like:
```
class ExpansionPosition {
   int Offset {get;}
   int Length {get;}
   ExpressionAst Expression {get;}
}
```
Or if backwards compatibility is a concern, a ```ReadOnlyList``` of ```{Offset+Length}``` could be added ```ExpandableStringExpressionAst```.

**Comment by lzybkr** [comment]
I don't see value in storing redundant data - the engine won't use that information at all.  I also want to reduce redundancy in what we store regarding position information - e.g. the filename is stored redundantly in every extent - it could be stored just once at the root of the ast.

I could see adding an api that returns a new Ast like you suggest - consider it a form of lowering, e.g. the api could return the format expression that way, so

```powershell
"`$foo = $foo"
```

would return an Ast that is equivalent, like:

```powershell
'$foo = {0}' -f $foo
```

I'm not sure what the extent of the StringLiteralAst in this new Ast  should be as it doesn't exactly correspond to anything in the source.

Note that this could be a messy api to implement - expandable strings can be recursive:

```powershell
@"
$("`$foo = $foo")
"@
```

**Comment by redoz** [comment]
Well my suggestion was just to include the constant string parts as Ast nodes in the ```NestedExpressions``` list, now it seems the parser discards them and only keeps the constant part in the ```Value``` (and ```FormatExpression```) properties. That means you could drop the ```FormatExpression``` and simply concat the results of each node in the list to get the expanded string. This seems like it should also handle recursion without issue and each node could have a valid and correct ```Extent``` as well.

**Comment by lzybkr** [comment]
One important design criteria was for there to be an obvious mapping from Ast node to something in the source script - at least as much as possible.

Why your proposal seems fine, I do think it goes against this design principle.

**Comment by redoz** [comment]
So just to ensure we're talking about the same thing, this is what I had in mind:
```
"`$foo = $foo"
```
 would get parsed into an `ExpandableStringExpressionAst` with an extent corresponding to the whole string (just as it is now). 
The `NestedExpressions` (or `Segments`?) property would contain two ast nodes, one
`ConstantStringExpressionAst` which would have an extent of  
```
`$foo = 
```
And a `VariableExpressionAst` with an extent of
```
$foo
```
If the design principle violation is that there are `ConstantStringExpressionAst` in the list, the  `Segments` property could contain a list of `ExpandableStringSegment` objects which are either `ConstantStringSegment` or `ExpressionSegment`. The former could have a `string` `Value` property, and the latter a `ExpressionAst` `Expression`  property.  That way the number of Ast nodes remain the same as they are today, just the structure of how they are represented would be easier to work with.

In the end I just don't want to have to do string parsing/manipulation in order to figure out what parts of the `Value` I need to join up with the evaluated Expressions as the parser already did this when it parsed the expression.

**Comment by lzybkr** [comment]
I understand. It's definitely one way of thinking about it, but I also think it might be surprising or confusing to many who work with the ast.

**Comment by redoz** [comment]
As long as it will be possible to round-trip this node without resorting to using its Extent or accessing internal properties I'll be happy. Hopefully you can find an internal representation that keeps everyone happy.
